### ISSUE TITLE
Symlinked module triggers recompilation

### ISSUE THREAD

**Issue by rdeits** [JuliaLang/julia#16007]
tl;dr: should we use `realpath()` when checking if a precompiled module matches its current path?

If I have a package which lives in, say, `~/foo/MyPackage/src/MyPackage`, and for which I've enabled precompilation, then I I can do:

``` bash
cd ~/foo/
julia -E "push(LOAD_PATH, \".\"); using MyPackage"
```

and the package will be precompiled exactly once. However, if I then do:

``` bash
mkdir ~/bar
ln -s /home/rdeits/foo/MyPackage /home/rdeits/bar/
cd bar
julia -E "push(LOAD_PATH, \".\"); using MyPackage"
```

then the package will be detected as stale and recompiled, even though the actual package files are unchanged and in the same (canonical) location. 

I believe that the fix for this is to use `realpath()` in [loading.jl](https://github.com/stevengj/julia/blob/4b24670cfa139a274a849a61836848ee56c038fe/base/loading.jl#L385) when checking if a module matches its cached path. I'm currently waiting for julia to build locally so I can verify this. 

But before I go too far, what is the desired behavior? Should a symlink to a module invalidate the cache? I think it shouldn't, but I'm not overly familiar with the guts of Julia. 

In case it's helpful, I'm actually interested in this because I've been playing around with [DeclarativePackages.jl](https://github.com/rened/DeclarativePackages.jl), which allows for more precise package versioning and non-global package installation. Unfortunately, the path checking in `stale_cachefile()` inside `loading.jl` interacts badly with the way DeclarativePackages is implemented. In essence, any time _any_ package version is changed with DeclarativePackages, _all_ modules get new paths, which triggers re-compilation of every single module. I'd love to avoid that if I can.

**Comment by gbaraldi** [comment]
This is fixed on master
